internal/server/singleprocess/service_trigger.go

```diff
@@ -71,3 +76,125 @@ func (s *service) ListTriggers(
 
 	return &pb.ListTriggerResponse{Triggers: result}, nil
 }
+
+func (s *service) RunTrigger(
+	ctx context.Context,
+	req *pb.RunTriggerRequest,
+) (*pb.RunTriggerResponse, error) {
+	if err := serverptypes.ValidateRunTriggerRequest(req); err != nil {
+		return nil, err
+	}
+	log := hclog.FromContext(ctx)
+
+	runTrigger, err := s.state.TriggerGet(req.Ref)
+	if err != nil {
+		return nil, err
+	}
+
+	log = log.With("run_trigger", runTrigger.Id)
+
+	log.Debug("building run trigger job")
+
+	// Build the job(s)
+	job := &pb.Job{
+		Workspace: runTrigger.Workspace,
+		Labels:    map[string]string{"trigger/id": runTrigger.Id},
+	}
+
+	// TODO(briancain): is there an easy way to convert this without the big switch
+	// so that we don't have to add new case statements any time a new operation is supported?
+	switch op := runTrigger.Operation.(type) {
+	case *pb.Trigger_Build:
+		job.Operation = &pb.Job_Build{Build: op.Build}
+	case *pb.Trigger_Push:
+		job.Operation = &pb.Job_Push{Push: op.Push}
+	case *pb.Trigger_Deploy:
+		job.Operation = &pb.Job_Deploy{Deploy: op.Deploy}
+	case *pb.Trigger_Destroy:
+		job.Operation = &pb.Job_Destroy{Destroy: op.Destroy}
+	case *pb.Trigger_Release:
+		job.Operation = &pb.Job_Release{Release: op.Release}
+	case *pb.Trigger_Up:
+		job.Operation = &pb.Job_Up{Up: op.Up}
+	case *pb.Trigger_Init:
+		job.Operation = &pb.Job_Init{Init: op.Init}
+	default:
+		return nil, status.Errorf(codes.Internal,
+			"trigger %q is configured with an unsupported operation %T", runTrigger.Id, op)
+	}
+
+	if len(req.VariableOverrides) > 0 {
+		log.Debug("variable overrides have been requested for trigger job")
+		job.Variables = req.VariableOverrides
+	}
+
+	// TODO(briancain): look up a target runner config at the project/app level and apply it to job requests
+	job.TargetRunner = &pb.Ref_Runner{Target: &pb.Ref_Runner_Any{}}
+
+	// generate job requests
+	var jobList []*pb.QueueJobRequest
+	var ids []string
+	if runTrigger.Application == nil {
+		// we're gonna queue multiple jobs for every application in a project
+		project, err := s.state.ProjectGet(runTrigger.Project)
+		if err != nil {
+			return nil, err
+		}
+
+		log.Debug("building multi-jobs for all apps in project", "project", project.Name)
+		for _, app := range project.Applications {
```

Comment:
Dunno if Rae's work added a way to queue a job for all applications, but this feels like a function that could live in the state interface instead.

Reply:
Good suggestion! I'll take a look, if not I can add it here or in a separate PR. It would be a nice addition.

Reply:
Got that added in this PR!